src/coprocessor/dag/expr/constant.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::borrow::Cow;
+
+use coprocessor::codec::Datum;
+use coprocessor::codec::mysql::{Decimal, Duration, Json, Time};
+use super::{Constant, Error, Result};
+
+#[inline]
+pub fn datum_as_int(d: &Datum) -> Result<Option<i64>> {
+    match *d {
+        Datum::I64(i) => Ok(Some(i)),
+        Datum::U64(u) => Ok(Some(u as i64)),
+        _ => Err(Error::Other("Can't eval_int from Constant")),
+    }
+}
+
+#[inline]
+pub fn datum_as_real(d: &Datum) -> Result<Option<f64>> {
+    match *d {
+        Datum::F64(f) => Ok(Some(f)),
+        _ => Err(Error::Other("Can't eval_real from Datum")),
+    }
+}
+
+#[inline]
+pub fn datum_as_decimal(d: &Datum) -> Result<Option<Cow<Decimal>>> {
+    match *d {
+        Datum::Dec(ref d) => Ok(Some(Cow::Borrowed(d))),
```

Comment:
I think we don't need clone that everytimes. For many cases, we just want to use the value, but not to modify it.